File: transposition.py
def encrypt(plaintext, key):
    col = len(key)
    row = len(plaintext) // col
    if len(plaintext) % col != 0:
        row += 1
    matrix = [['' for _ in range(col)] for _ in range(row)]
    for i in range(len(plaintext)):
        matrix[i // col][i % col] = plaintext[i]
    
    key_order = sorted([(key[i], i) for i in range(len(key))])
    
    cipher = ''
    for _, index in key_order:
        for r in range(row):
            cipher += matrix[r][index] if matrix[r][index] else 'Z'
    
    return cipher

def decrypt(cipher, key):
    col = len(key)
    row = len(cipher) // col
    
    key_order = sorted([(key[i], i) for i in range(len(key))])
    index_order = [i[1] for i in key_order]
    
    matrix = [['' for _ in range(col)] for _ in range(row)]
    idx = 0
    for i in index_order:
        for j in range(row):
            matrix[j][i] = cipher[idx]
            idx += 1
    
    plaintext = ''.join(''.join(row) for row in matrix).rstrip('Z')
    return plaintext

if __name__ == "__main__":
    plaintext = input("Enter plaintext: ").replace(" ", "").upper()
    key = input("Enter key: ").upper()
    cipher_text = encrypt(plaintext, key)
    print(f"Encrypted: {cipher_text}")
    decrypted_text = decrypt(cipher_text, key)
    print(f"Decrypted: {decrypted_text}")

File: transposition.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

// Function to encrypt using columnar transposition
string encrypt(const string& plaintext, const string& key) {
    int keyLength = key.length();
    int plaintextLength = plaintext.length();

    // Determine the number of rows needed
    int numRows = (plaintextLength + keyLength - 1) / keyLength;

    // Create a matrix to hold the plaintext characters
    vector<vector<char>> matrix(numRows, vector<char>(keyLength, ' ')); // Initialize with spaces

    // Fill the matrix with the plaintext
    int k = 0;
    for (int i = 0; i < numRows; ++i) {
        for (int j = 0; j < keyLength; ++j) {
            if (k < plaintextLength) {
                matrix[i][j] = plaintext[k++];
            }
        }
    }

    // Sort the key to determine the column order
    string sortedKey = key;
    sort(sortedKey.begin(), sortedKey.end());

    // Read the matrix column by column based on the sorted key
    string ciphertext = "";
    for (char c : sortedKey) {
        int colIndex = key.find(c);
        for (int i = 0; i < numRows; ++i) {
            ciphertext += matrix[i][colIndex];
        }
        key[colIndex] = '*'; // Mark the column as used to handle duplicate key characters
    }

    return ciphertext;
}

// Function to decrypt using columnar transposition
string decrypt(const string& ciphertext, const string& key) {
    int keyLength = key.length();
    int ciphertextLength = ciphertext.length();
    int numRows = (ciphertextLength + keyLength - 1) / keyLength;

    // Create a matrix to hold the ciphertext characters
    vector<vector<char>> matrix(numRows, vector<char>(keyLength, ' '));

    // Sort the key to determine the column order
    string sortedKey = key;
    sort(sortedKey.begin(), sortedKey.end());

    // Fill the matrix column by column based on the sorted key
    int k = 0;
    string originalKey = key;  // Store the original key for index lookup
    for (char c : sortedKey) {
        int colIndex = originalKey.find(c);

        for (int i = 0; i < numRows; ++i) {
            if (k < ciphertextLength) {
                matrix[i][colIndex] = ciphertext[k++];
            }
        }
        originalKey[colIndex] = '*';  // Mark the column as used to handle duplicate key characters
    }

    // Read the matrix row by row to get the plaintext
    string plaintext = "";
    for (int i = 0; i < numRows; ++i) {
        for (int j = 0; j < keyLength; ++j) {
            if (matrix[i][j] != ' ') {
                plaintext += matrix[i][j];
            }
        }
    }

    return plaintext;
}

int main() {
    string plaintext, key;

    cout << "Enter the plaintext: ";
    getline(cin, plaintext);

    cout << "Enter the key (e.g., 4312): ";
    getline(cin, key);

    string ciphertext = encrypt(plaintext, key);
    cout << "Ciphertext: " << ciphertext << endl;

    string decryptedText = decrypt(ciphertext, key);
    cout << "Decrypted text: " << decryptedText << endl;

    return 0;
}